### ISSUE TITLE
Running the game

### ISSUE THREAD

**Issue by mardy** [xtreme8000/CavEX#15]
Hi! Sorry for the silly question, but I'm not able to run the game: I built both the Wii and the Linux version, and I get the following issues:

**Linux version**

The game starts, it recognized my two save folders, but does not react to input. Mouse and keyboard presses have no effect. The  game does not seem to be using the CPU, but it cannot be closed by closing the window -- only by pressing Ctrl+C on the terminal.

**Wii version**

I created a directory containing the DOL file (renamed to boot.dol), plus the assets, the save files, and the config, icon and meta files from the source tree. Then I try to open the DOL file with dolphin, but it immediately asks me if I want to stop the current emulation.

**Comment by mardy** [comment]
> The game starts, it recognized my two save folders, but does not react to input.

OK, this was due to me missing the config.json file. Moving the `config_pc.json` file into the build directory (and renaming it to `config.json`) did the trick. :-)

For the Wii version I'm still trying to understand what happens.

**Comment by mardy** [comment]
> I created a directory containing the DOL file (renamed to boot.dol), plus the assets, the save files, and the config, icon and meta files from the source tree. Then I try to open the DOL file with dolphin, but it immediately asks me if I want to stop the current emulation.

The same happens on a real Wii as well: after a couple of seconds (during which only a black screen is seen) the game quits and I get back to the homebrew channel app.

I will try building the `v0.2.1` tag to see if it might be due to a later change.

**Comment by SmollestBug** [comment]
I had this same issue explained in [Issue #10 comment](https://github.com/xtreme8000/CavEX/issues/10#issuecomment-1673119809) and you just need a couple older install files for it to compile properly. There are a bunch of older installs here, https://gbatemp.net/threads/collection-of-old-devkitpro-versions.526377/ and from what I've tried you can just use the newest versions dated August 19th, 2022 of DevkitPPC and libogc.

The compiled build can crash decently often, and it probably is because of the cursed install used to build it. Unless Bytebit says otherwise.

**Comment by xtreme8000** [comment]
@SmollestBug what kind of crash? 😟 It should be very stable on Wii.

**Comment by SmollestBug** [comment]
@xtreme8000 Basically just between 5-20 minutes of flying around a world it'll just inexplicably freeze, it's also frozen and then unfrozen a couple seconds later before, but then enters a state where it stops loading chunks from the world file.

**Comment by mardy** [comment]
> I had this same issue explained in [Issue #10 comment](https://github.com/xtreme8000/CavEX/issues/10#issuecomment-1673119809) and you just need a couple older install files for it to compile properly. There are a bunch of older installs here, https://gbatemp.net/threads/collection-of-old-devkitpro-versions.526377/ and from what I've tried you can just use the newest versions dated August 19th, 2022 of DevkitPPC and libogc.

Mmm... my devkitPPC version is r41, which is exactly the version you suggest:
```
$ /opt/devkitpro/devkitPPC/bin/powerpc-eabi-gcc --version
powerpc-eabi-gcc (devkitPPC release 41) 12.1.0
Copyright (C) 2022 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

My libogc seems to be actually older, it's dated July 2nd, 2021 and looking at the strings inside the binary I found this
```
GCC: (devkitPPC release 39) 11.1.0
```
I'll try upgrading libogc, then, to the version from August 2022, and report back. :-)

**Comment by mardy** [comment]
Nope, no luck, it makes no difference.

I've also tried to checkout an older commit:
- `v0.2.1` does not build, due to missing `textures.h`
- commit 685d904bb4534684681ae952da69da1595e3b199 builds and actually even runs, but the screen stays black (but at least the game does not seem to crash, dolphin reports it as working)

Just to make sure that it's not an issue with the emulator, I took the `boot.dol` from the v0.2.1 release and moved it over my compiled one, and indeed that works. @SmollestBug, what commit are you building from?

It might also be that it's due to changes in the versions of the dependent libraries. I'm using:

- cNBT commit 3ff982eff62b01f7a67079a365b8fbf8a4b47423
- cglm commit f496146bce3ac33e6a13a65baeb30d36d1e9bbfb
- lodepng commit c18b949b71f45e78b1f9a28c5d458bce0da505d6
- m-lib commit 469239adcd1c37f09e208b7c09fe7f749de84892
- parson commit fd02ea0c56e14cce9d4114c06b5a0d5e427fe4d8

**Comment by SmollestBug** [comment]
@mardy Maybe try a fresh install of latest devkitpro, and then downgrade devkitPPC and libogc. This is how I did it.

**Comment by mardy** [comment]
Still no luck. I also tried the following things, but none helped:
- Switch optimizations off (-O0)
- Comment out the calls to `usleep()`

The latter seems to prove that this is not the same issue mentioned in issue #10.

I then tried to run a git bisect, to find out what is the last commit where the game does not exit immediately, and that's commit ddf6827254020e81057fa7a5d6b5ae414be01699; meaning, that the first commit where things broke (well, I get a black screen on the "working" one as well :-)) is commit 511e695e49f0fd87cca1983cdfabdcbc22ffd918.

I then took the latest master branch, added a call to `PAD_Init()` in `input_init()`, and now the game does not exit immediately, but stays running (still, only a black screen is shown). But I have no idea if this is part of the problem, or if it's irrelevant. @xtreme8000, do you have any idea of what could be wrong? Or maybe a suggestion to debug this (i'm new to Wii development, so I don't know if there's an equivalent of printf() that I could use...)?

**Comment by xtreme8000** [comment]
Obviously there is no `printf()` on Wii, so in severe cases I just used `exit(0)` and tried it at different places (as suggested first [here](http://wiibrew.org/wiki/Debugging#Alternative_Method_3)). In Dolphin you can easily see that it reached `exit`, because it will show a prompt.

Make sure to not call `PAD_Init()` anywhere, because it makes `exit` hang indefinitely for some unknown reason.

**Comment by mardy** [comment]
> Obviously there is no `printf()` on Wii, so in severe cases I just used `exit(0)` and tried it at different places (as suggested first [here](http://wiibrew.org/wiki/Debugging#Alternative_Method_3)).

Thanks for the advice! Since for me the game exits immediately, whereas with PAD_Input() it hangs, I debugged it using the latter, and checked when it was *not* exiting. :-)

It looks like a deployment issue: the game quits because the textures cannot be loaded (`tex_read()` returns NULL when trying to load the `terrain.png` atlas). The weird thing is that when I run dolphin under `strace` I don't see any attempt to open the `assets` directory, and even more strange is the fact that the binary `boot.dol` that you published for v0.2.1 works fine, yet I don't see that it's opening the assets directory either (at least according to strace). What's the trick here? Maybe it had the textures built into the binary back then?

**Comment by xtreme8000** [comment]
Yes, the old v0.2.1 release had all assets inside the binary.

You probably don't see Dolphin opening a directory, because you need to use an emulated SDcard, from which it reads any files from. In Dolphin's Wii settings, you can set the location of a raw image, for me the default was `~/.local/share/dolphin-emu/Load/WiiSD.raw`.
That can be mounted e.g. `sudo mount -t vfat -o user,uid=1000,gid=1000 ~/.local/share/dolphin-emu/Load/WiiSD.raw ~/sdcard/` (format first).

**Comment by mardy** [comment]
Oops, I wrote too soon: as a matter of fact, dolphin was still loading the dol file from the official release, so it turns out I didn't fix anything :-(

**Comment by SmollestBug** [comment]
I don't know if this could help narrow it down, I don't have much experience with this. But what is the file size of the .dol file you're getting? I noticed that depending on what version devkitpro was on that it would compile different size .dol files.

Up to date version gives a 875kb .dol
Purposely downgraded version I suggested gives a 905kb .dol

I'd guess it changes quite a bit depending on how old the version you're using is. But I could assume if something is going wrong then it could be outputting some flimsy broken file.

**Comment by mardy** [comment]
> I don't know if this could help narrow it down, I don't have much experience with this. But what is the file size of the .dol file you're getting? I noticed that depending on what version devkitpro was on that it would compile different size .dol files.

It's 932288 bytes. But now I'm convinced that the issue I'm seeing is just that dolphin cannot find the assets: I've just tried on a real Wii, and the game works just fine: I first tried with usleep() commented out, and then with the usleep() call, and it works in both cases. The reason why it didn't work the first time I tried on the wii, is that I didn't copy the assets, since I installed it over an existing installation of cavex and I thought that the assets would be the same (wrong! :-) ).

So, this is not a compiler issue, it's just that I'm unable to get dolphin emulate the SD card properly (and developing using Dolphin would be much easier for me). I'll give it a few more attempts. :-)
